HDI version references break now that there's 0.X.X versions

## Symptoms
See the comment below, with the full output

## Steps to reproduce
`cargo init` a new cargo crate
set `hdk = "0.0.152"` as a dependency
run `cargo build`. 

## Analysis

This hasn't previously been an issue when using `0.0.X` dependencies because every version was treated as incompatible with other 0.0.X versions, due to being the left-most non-zero number.

https://github.com/holochain/holochain/blob/f0264303503df5b872de03d4f991822b34689a9b/crates/hdk/Cargo.toml#L30

`cargo` actually resolves this to `0.1.3` when 0.1.3 has been released, or 0.1.4 when that's released, etc. 

I had already dealt with the same or similar thing in holochain-open-dev/dna-auth-resolver and it's an issue that comes up when you change to using `minor` versions in SemVer from just using `patch` versions. 

I had to analyze this confusing situation we had, and found this: 
https://doc.rust-lang.org/cargo/reference/resolver.html#semver-compatibility

I've done some digging... in the cargo dep resolution book: 
"Versions are considered compatible if their left-most non-zero major/minor/patch component is the same. For example, 1.0.3 and 1.1.0 are considered compatible, and thus it should be safe to update from the older release to the newer one. However, an update from 1.1.0 to 2.0.0 would not be allowed to be made automatically. This convention also applies to versions with leading zeros. For example, 0.1.0 and 0.1.2 are compatible, but 0.1.0 and 0.2.0 are not. Similarly, 0.0.1 and 0.0.2 are not compatible."
That part at the end is key. 
"Similarly, 0.0.1 and 0.0.2 are not compatible."
It also says: 
"If multiple packages have a common dependency with semver-incompatible versions, then Cargo will allow this, but will build two separate copies of the dependency. "

## Fix Suggestion

switch to `=0.1.2` or equivalent


## Update

I figured out that in the short run we can workaround this with this command:
```
cargo update -p hdi --precise 0.1.2
```
but that will not be easy or intuitive for anyone to figure out

Here is the full: 
```
MacBook-Pro:test-hdk connor$ cargo build
    Updating crates.io index
  Downloaded proc-macro2 v1.0.44
  Downloaded thiserror v1.0.37
  Downloaded thiserror-impl v1.0.37
  Downloaded serde v1.0.145
  Downloaded iana-time-zone v0.1.50
  Downloaded syn v1.0.101
  Downloaded kitsune_p2p_dht_arc v0.0.16
  Downloaded holo_hash v0.0.33
  Downloaded holochain_integrity_types v0.0.20
  Downloaded hdi v0.1.3
  Downloaded hdk_derive v0.0.50
  Downloaded serde_derive v1.0.145
  Downloaded semver v1.0.14
  Downloaded lock_api v0.4.9
  Downloaded 14 crates (649.8 KB) in 0.62s
   Compiling proc-macro2 v1.0.44
   Compiling unicode-ident v1.0.4
   Compiling quote v1.0.21
   Compiling syn v1.0.101
   Compiling autocfg v1.1.0
   Compiling libc v0.2.133
   Compiling cfg-if v1.0.0
   Compiling serde_derive v1.0.145
   Compiling serde v1.0.145
   Compiling version_check v0.9.4
   Compiling once_cell v1.15.0
   Compiling thiserror v1.0.37
   Compiling fnv v1.0.7
   Compiling ident_case v1.0.1
   Compiling memchr v2.5.0
   Compiling serde_json v1.0.85
   Compiling paste v1.0.5
   Compiling byteorder v1.4.3
   Compiling ryu v1.0.11
   Compiling itoa v1.0.3
   Compiling cc v1.0.73
   Compiling bytecheck v0.6.9
   Compiling bit-vec v0.6.3
   Compiling convert_case v0.4.0
   Compiling rustversion v1.0.9
   Compiling trilean v1.1.0
   Compiling rend v0.3.6
   Compiling smallvec v1.9.0
   Compiling rkyv v0.7.39
   Compiling log v0.4.17
   Compiling target-lexicon v0.12.4
   Compiling gimli v0.26.2
   Compiling adler v1.0.2
   Compiling seahash v4.1.0
   Compiling bitflags v1.3.2
   Compiling crossbeam-utils v0.8.11
   Compiling rustc-demangle v0.1.21
   Compiling more-asserts v0.2.2
   Compiling cranelift-entity v0.76.0
   Compiling scopeguard v1.1.0
   Compiling wasmparser v0.78.2
   Compiling cranelift-codegen-shared v0.76.0
   Compiling rayon-core v1.9.3
   Compiling leb128 v0.2.5
   Compiling hashbrown v0.12.3
   Compiling crc32fast v1.3.2
   Compiling pin-project-lite v0.2.9
   Compiling either v1.8.0
   Compiling core-foundation-sys v0.8.3
   Compiling rustc-hash v1.1.0
   Compiling fallible-iterator v0.2.0
   Compiling stable_deref_trait v1.2.0
   Compiling lazy_static v1.4.0
   Compiling subtle v2.4.1
   Compiling fastrand v1.8.0
   Compiling remove_dir_all v0.5.3
   Compiling unicode-width v0.1.10
   Compiling wasmer-compiler-cranelift v2.2.0
   Compiling strsim v0.10.0
   Compiling parking_lot_core v0.8.5
   Compiling heck v0.4.0
   Compiling libloading v0.7.3
   Compiling instant v0.1.12
   Compiling tracing-core v0.1.29
   Compiling ahash v0.7.6
   Compiling proc-macro-error-attr v1.0.4
   Compiling proc-macro-error v1.0.4
   Compiling num-traits v0.2.15
   Compiling indexmap v1.9.1
   Compiling num-integer v0.1.45
   Compiling memoffset v0.6.5
   Compiling crossbeam-epoch v0.9.10
   Compiling rayon v1.5.3
   Compiling lock_api v0.4.9
   Compiling bit-set v0.5.3
   Compiling miniz_oxide v0.5.4
   Compiling wasm-encoder v0.17.0
   Compiling cranelift-codegen-meta v0.76.0
   Compiling cranelift-bforest v0.76.0
   Compiling backtrace v0.3.66
   Compiling wasmer-vm v2.2.0
   Compiling addr2line v0.17.0
   Compiling regalloc v0.0.31
   Compiling getrandom v0.2.7
   Compiling mach v0.3.2
   Compiling num_cpus v1.13.1
   Compiling memmap2 v0.5.7
   Compiling tempfile v3.3.0
   Compiling which v4.3.0
   Compiling time v0.1.44
   Compiling iana-time-zone v0.1.50
   Compiling object v0.29.0
   Compiling wast v47.0.0
   Compiling crossbeam-channel v0.5.6
   Compiling parking_lot v0.11.2
   Compiling region v3.0.0
   Compiling crossbeam-deque v0.8.2
   Compiling hashbrown v0.11.2
   Compiling rmp v0.8.11
   Compiling gcollections v1.5.0
   Compiling intervallum v1.4.0
   Compiling cranelift-codegen v0.76.0
   Compiling darling_core v0.13.4
   Compiling darling_core v0.14.1
   Compiling wat v1.0.49
   Compiling thiserror-impl v1.0.37
   Compiling ptr_meta_derive v0.1.4
   Compiling derive_more v0.99.17
   Compiling holochain_serialized_bytes_derive v0.0.51
   Compiling bytecheck_derive v0.6.9
   Compiling loupe-derive v0.1.3
   Compiling rkyv_derive v0.7.39
   Compiling enum-iterator-derive v0.7.0
   Compiling tracing-attributes v0.1.22
   Compiling wasmer-derive v2.2.0
   Compiling ptr_meta v0.1.4
   Compiling darling_macro v0.14.1
   Compiling darling_macro v0.13.4
   Compiling enum-iterator v0.7.0
   Compiling darling v0.14.1
   Compiling darling v0.13.4
   Compiling enumset_derive v0.6.0
   Compiling tracing v0.1.36
   Compiling enumset v1.0.11
   Compiling serde_bytes v0.11.7
   Compiling serde-transcode v1.1.1
   Compiling rmp-serde v0.15.5
   Compiling kitsune_p2p_dht_arc v0.0.16
   Compiling kitsune_p2p_dht_arc v0.0.15
   Compiling kitsune_p2p_timestamp v0.0.14
   Compiling chrono v0.4.22
   Compiling loupe v0.1.3
   Compiling gimli v0.25.0
   Compiling object v0.28.4
   Compiling wasmer-types v2.2.0
   Compiling holochain_serialized_bytes v0.0.51
   Compiling holo_hash v0.0.33
   Compiling holo_hash v0.0.32
   Compiling holochain_integrity_types v0.0.19
   Compiling holochain_integrity_types v0.0.20
   Compiling wasmer-compiler v2.2.0
   Compiling wasmer-engine v2.2.0
   Compiling wasmer-object v2.2.0
   Compiling wasmer-engine-dylib v2.2.0
   Compiling wasmer-engine-universal v2.2.0
   Compiling hdk_derive v0.0.49
   Compiling hdk_derive v0.0.50
   Compiling cranelift-frontend v0.76.0
   Compiling wasmer v2.2.0
   Compiling holochain_wasmer_common v0.0.80
   Compiling holochain_zome_types v0.0.48
   Compiling holochain_wasmer_guest v0.0.80
   Compiling hdi v0.1.3
   Compiling hdk v0.0.152
error[E0053]: method `verify_signature` has an incompatible type for trait
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/hdk.rs:208:51
    |
208 |     fn verify_signature(&self, _verify_signature: VerifySignature) -> ExternResult<bool> {
    |                                                   ^^^^^^^^^^^^^^^
    |                                                   |
    |                                                   expected struct `hdi::prelude::VerifySignature`, found struct `holochain_zome_types::VerifySignature`
    |                                                   help: change the parameter type to match the trait: `hdi::prelude::VerifySignature`
    |
    = note: expected fn pointer `fn(&ErrHdk, hdi::prelude::VerifySignature) -> Result<_, _>`
               found fn pointer `fn(&ErrHdk, holochain_zome_types::VerifySignature) -> Result<_, _>`

error[E0053]: method `hash` has an incompatible type for trait
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/hdk.rs:212:33
    |
212 |     fn hash(&self, _hash_input: HashInput) -> ExternResult<HashOutput> {
    |                                 ^^^^^^^^^
    |                                 |
    |                                 expected enum `hdi::prelude::HashInput`, found enum `holochain_zome_types::HashInput`
    |                                 help: change the parameter type to match the trait: `hdi::prelude::HashInput`
    |
    = note: expected fn pointer `fn(&ErrHdk, hdi::prelude::HashInput) -> Result<hdi::prelude::HashOutput, _>`
               found fn pointer `fn(&ErrHdk, holochain_zome_types::HashInput) -> Result<holochain_zome_types::HashOutput, _>`

error[E0053]: method `must_get_entry` has an incompatible type for trait
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/hdk.rs:218:32
    |
218 |         _must_get_entry_input: MustGetEntryInput,
    |                                ^^^^^^^^^^^^^^^^^
    |                                |
    |                                expected struct `hdi::prelude::MustGetEntryInput`, found struct `holochain_zome_types::MustGetEntryInput`
    |                                help: change the parameter type to match the trait: `hdi::prelude::MustGetEntryInput`
    |
    = note: expected fn pointer `fn(&ErrHdk, hdi::prelude::MustGetEntryInput) -> Result<hdi::prelude::HoloHashed<hdi::prelude::Entry>, _>`
               found fn pointer `fn(&ErrHdk, holochain_zome_types::MustGetEntryInput) -> Result<holo_hash::HoloHashed<holochain_zome_types::Entry>, _>`

error[E0053]: method `must_get_action` has an incompatible type for trait
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/hdk.rs:225:33
    |
225 |         _must_get_action_input: MustGetActionInput,
    |                                 ^^^^^^^^^^^^^^^^^^
    |                                 |
    |                                 expected struct `hdi::prelude::MustGetActionInput`, found struct `holochain_zome_types::MustGetActionInput`
    |                                 help: change the parameter type to match the trait: `hdi::prelude::MustGetActionInput`
    |
    = note: expected fn pointer `fn(&ErrHdk, hdi::prelude::MustGetActionInput) -> Result<hdi::prelude::SignedHashed<hdi::prelude::Action>, _>`
               found fn pointer `fn(&ErrHdk, holochain_zome_types::MustGetActionInput) -> Result<holochain_zome_types::SignedHashed<holochain_zome_types::Action>, _>`

error[E0053]: method `must_get_valid_record` has an incompatible type for trait
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/hdk.rs:232:39
    |
232 |         _must_get_valid_record_input: MustGetValidRecordInput,
    |                                       ^^^^^^^^^^^^^^^^^^^^^^^
    |                                       |
    |                                       expected struct `hdi::prelude::MustGetValidRecordInput`, found struct `holochain_zome_types::MustGetValidRecordInput`
    |                                       help: change the parameter type to match the trait: `hdi::prelude::MustGetValidRecordInput`
    |
    = note: expected fn pointer `fn(&ErrHdk, hdi::prelude::MustGetValidRecordInput) -> Result<hdi::prelude::Record, _>`
               found fn pointer `fn(&ErrHdk, holochain_zome_types::MustGetValidRecordInput) -> Result<holochain_zome_types::Record, _>`

error[E0053]: method `must_get_agent_activity` has an incompatible type for trait
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/hdk.rs:239:12
    |
239 |         _: MustGetAgentActivityInput,
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^
    |            |
    |            expected struct `hdi::prelude::MustGetAgentActivityInput`, found struct `holochain_zome_types::MustGetAgentActivityInput`
    |            help: change the parameter type to match the trait: `hdi::prelude::MustGetAgentActivityInput`
    |
    = note: expected fn pointer `fn(&ErrHdk, hdi::prelude::MustGetAgentActivityInput) -> Result<Vec<hdi::prelude::RegisterAgentActivity>, _>`
               found fn pointer `fn(&ErrHdk, holochain_zome_types::MustGetAgentActivityInput) -> Result<Vec<holochain_zome_types::RegisterAgentActivity>, _>`

error[E0053]: method `dna_info` has an incompatible type for trait
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/hdk.rs:244:48
    |
244 |     fn dna_info(&self, _dna_info_input: ()) -> ExternResult<DnaInfo> {
    |                                                ^^^^^^^^^^^^^^^^^^^^^
    |                                                |
    |                                                expected struct `hdi::prelude::DnaInfo`, found struct `holochain_zome_types::DnaInfo`
    |                                                help: change the output type to match the trait: `Result<hdi::prelude::DnaInfo, holochain_wasmer_guest::WasmError>`
    |
    = note: expected fn pointer `fn(&ErrHdk, ()) -> Result<hdi::prelude::DnaInfo, _>`
               found fn pointer `fn(&ErrHdk, ()) -> Result<holochain_zome_types::DnaInfo, _>`

error[E0053]: method `zome_info` has an incompatible type for trait
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/hdk.rs:248:50
    |
248 |     fn zome_info(&self, _zome_info_input: ()) -> ExternResult<ZomeInfo> {
    |                                                  ^^^^^^^^^^^^^^^^^^^^^^
    |                                                  |
    |                                                  expected struct `hdi::prelude::ZomeInfo`, found struct `holochain_zome_types::ZomeInfo`
    |                                                  help: change the output type to match the trait: `Result<hdi::prelude::ZomeInfo, holochain_wasmer_guest::WasmError>`
    |
    = note: expected fn pointer `fn(&ErrHdk, ()) -> Result<hdi::prelude::ZomeInfo, _>`
               found fn pointer `fn(&ErrHdk, ()) -> Result<holochain_zome_types::ZomeInfo, _>`

error[E0053]: method `trace` has an incompatible type for trait
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/hdk.rs:252:24
    |
252 |     fn trace(&self, _: TraceMsg) -> ExternResult<()> {
    |                        ^^^^^^^^
    |                        |
    |                        expected struct `hdi::prelude::TraceMsg`, found struct `holochain_zome_types::TraceMsg`
    |                        help: change the parameter type to match the trait: `hdi::prelude::TraceMsg`
    |
    = note: expected fn pointer `fn(&ErrHdk, hdi::prelude::TraceMsg) -> Result<_, _>`
               found fn pointer `fn(&ErrHdk, holochain_zome_types::TraceMsg) -> Result<_, _>`

error[E0053]: method `x_salsa20_poly1305_decrypt` has an incompatible type for trait
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/hdk.rs:258:38
    |
258 |         _x_salsa20_poly1305_decrypt: XSalsa20Poly1305Decrypt,
    |                                      ^^^^^^^^^^^^^^^^^^^^^^^
    |                                      |
    |                                      expected struct `hdi::prelude::XSalsa20Poly1305Decrypt`, found struct `holochain_zome_types::XSalsa20Poly1305Decrypt`
    |                                      help: change the parameter type to match the trait: `hdi::prelude::XSalsa20Poly1305Decrypt`
    |
    = note: expected fn pointer `fn(&ErrHdk, hdi::prelude::XSalsa20Poly1305Decrypt) -> Result<std::option::Option<hdi::prelude::XSalsa20Poly1305Data>, _>`
               found fn pointer `fn(&ErrHdk, holochain_zome_types::XSalsa20Poly1305Decrypt) -> Result<std::option::Option<holochain_zome_types::XSalsa20Poly1305Data>, _>`

error[E0053]: method `x_25519_x_salsa20_poly1305_decrypt` has an incompatible type for trait
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/hdk.rs:265:46
    |
265 |         _x_25519_x_salsa20_poly1305_decrypt: X25519XSalsa20Poly1305Decrypt,
    |                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |                                              |
    |                                              expected struct `hdi::prelude::X25519XSalsa20Poly1305Decrypt`, found struct `holochain_zome_types::X25519XSalsa20Poly1305Decrypt`
    |                                              help: change the parameter type to match the trait: `hdi::prelude::X25519XSalsa20Poly1305Decrypt`
    |
    = note: expected fn pointer `fn(&ErrHdk, hdi::prelude::X25519XSalsa20Poly1305Decrypt) -> Result<std::option::Option<hdi::prelude::XSalsa20Poly1305Data>, _>`
               found fn pointer `fn(&ErrHdk, holochain_zome_types::X25519XSalsa20Poly1305Decrypt) -> Result<std::option::Option<holochain_zome_types::XSalsa20Poly1305Data>, _>`

error[E0277]: the trait bound `hdi::prelude::Entry: From<holochain_zome_types::Entry>` is not satisfied
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/hash_path/path.rs:15:5
    |
15  |     hash_entry(Entry::App(
    |     ^^^^^^^^^^ the trait `From<holochain_zome_types::Entry>` is not implemented for `hdi::prelude::Entry`
    |
    = help: the trait `From<hdi::prelude::HoloHashed<hdi::prelude::Entry>>` is implemented for `hdi::prelude::Entry`
    = note: required because of the requirements on the impl of `Into<hdi::prelude::Entry>` for `holochain_zome_types::Entry`
    = note: required because of the requirements on the impl of `holochain_wasmer_guest::TryFrom<holochain_zome_types::Entry>` for `hdi::prelude::Entry`
note: required by a bound in `hdi::hash::hash_entry`
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdi-0.1.3/src/hash.rs:198:12
    |
198 |     Entry: TryFrom<I, Error = E>,
    |            ^^^^^^^^^^^^^^^^^^^^^ required by this bound in `hdi::hash::hash_entry`

error[E0277]: the trait bound `holo_hash::HoloHash<holo_hash::hash_type::AnyLinkable>: From<hdi::prelude::HoloHash<hdi::prelude::holo_hash::hash_type::Entry>>` is not satisfied
  --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/hash_path/path.rs:15:5
   |
15 | /     hash_entry(Entry::App(
16 | |         AppEntryBytes::try_from(SerializedBytes::from(UnsafeBytes::from(ROOT.to_vec())))
17 | |             .expect("This cannot fail as it's under the max entry bytes"),
18 | |     ))
19 | |     .map(Into::into)
   | |____________________^ the trait `From<hdi::prelude::HoloHash<hdi::prelude::holo_hash::hash_type::Entry>>` is not implemented for `holo_hash::HoloHash<holo_hash::hash_type::AnyLinkable>`
   |
   = help: the following other types implement trait `From<T>`:
             <holo_hash::HoloHash<holo_hash::hash_type::Action> as From<holo_hash::HoloHash<holo_hash::hash_type::AnyDht>>>
             <holo_hash::HoloHash<holo_hash::hash_type::Action> as From<holo_hash::HoloHash<holo_hash::hash_type::AnyLinkable>>>
             <holo_hash::HoloHash<holo_hash::hash_type::Agent> as From<holo_hash::HoloHash<holo_hash::hash_type::Entry>>>
             <holo_hash::HoloHash<holo_hash::hash_type::AnyDht> as From<holo_hash::HoloHash<holo_hash::hash_type::Action>>>
             <holo_hash::HoloHash<holo_hash::hash_type::AnyDht> as From<holo_hash::HoloHash<holo_hash::hash_type::Agent>>>
             <holo_hash::HoloHash<holo_hash::hash_type::AnyDht> as From<holo_hash::HoloHash<holo_hash::hash_type::AnyLinkable>>>
             <holo_hash::HoloHash<holo_hash::hash_type::AnyDht> as From<holo_hash::HoloHash<holo_hash::hash_type::Entry>>>
             <holo_hash::HoloHash<holo_hash::hash_type::AnyLinkable> as From<holo_hash::HoloHash<holo_hash::hash_type::Action>>>
           and 7 others
   = note: required because of the requirements on the impl of `Into<holo_hash::HoloHash<holo_hash::hash_type::AnyLinkable>>` for `hdi::prelude::HoloHash<hdi::prelude::holo_hash::hash_type::Entry>`

error[E0277]: the trait bound `holo_hash::HoloHash<holo_hash::hash_type::AnyLinkable>: From<hdi::prelude::HoloHash<hdi::prelude::holo_hash::hash_type::Entry>>` is not satisfied
  --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/hash_path/path.rs:19:10
   |
19 |     .map(Into::into)
   |          ^^^^^^^^^^ the trait `From<hdi::prelude::HoloHash<hdi::prelude::holo_hash::hash_type::Entry>>` is not implemented for `holo_hash::HoloHash<holo_hash::hash_type::AnyLinkable>`
   |
   = help: the following other types implement trait `From<T>`:
             <holo_hash::HoloHash<holo_hash::hash_type::Action> as From<holo_hash::HoloHash<holo_hash::hash_type::AnyDht>>>
             <holo_hash::HoloHash<holo_hash::hash_type::Action> as From<holo_hash::HoloHash<holo_hash::hash_type::AnyLinkable>>>
             <holo_hash::HoloHash<holo_hash::hash_type::Agent> as From<holo_hash::HoloHash<holo_hash::hash_type::Entry>>>
             <holo_hash::HoloHash<holo_hash::hash_type::AnyDht> as From<holo_hash::HoloHash<holo_hash::hash_type::Action>>>
             <holo_hash::HoloHash<holo_hash::hash_type::AnyDht> as From<holo_hash::HoloHash<holo_hash::hash_type::Agent>>>
             <holo_hash::HoloHash<holo_hash::hash_type::AnyDht> as From<holo_hash::HoloHash<holo_hash::hash_type::AnyLinkable>>>
             <holo_hash::HoloHash<holo_hash::hash_type::AnyDht> as From<holo_hash::HoloHash<holo_hash::hash_type::Entry>>>
             <holo_hash::HoloHash<holo_hash::hash_type::AnyLinkable> as From<holo_hash::HoloHash<holo_hash::hash_type::Action>>>
           and 7 others
   = note: required because of the requirements on the impl of `Into<holo_hash::HoloHash<holo_hash::hash_type::AnyLinkable>>` for `hdi::prelude::HoloHash<hdi::prelude::holo_hash::hash_type::Entry>`

error[E0277]: the trait bound `hdi::prelude::Entry: From<holochain_zome_types::Entry>` is not satisfied
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/hash_path/path.rs:279:9
    |
279 |         hash_entry(Entry::App(AppEntryBytes(
    |         ^^^^^^^^^^ the trait `From<holochain_zome_types::Entry>` is not implemented for `hdi::prelude::Entry`
    |
    = help: the trait `From<hdi::prelude::HoloHashed<hdi::prelude::Entry>>` is implemented for `hdi::prelude::Entry`
    = note: required because of the requirements on the impl of `Into<hdi::prelude::Entry>` for `holochain_zome_types::Entry`
    = note: required because of the requirements on the impl of `holochain_wasmer_guest::TryFrom<holochain_zome_types::Entry>` for `hdi::prelude::Entry`
note: required by a bound in `hdi::hash::hash_entry`
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdi-0.1.3/src/hash.rs:198:12
    |
198 |     Entry: TryFrom<I, Error = E>,
    |            ^^^^^^^^^^^^^^^^^^^^^ required by this bound in `hdi::hash::hash_entry`

error[E0308]: mismatched types
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/hash_path/path.rs:279:9
    |
278 |       pub fn path_entry_hash(&self) -> ExternResult<holo_hash::EntryHash> {
    |                                        ---------------------------------- expected `Result<holo_hash::HoloHash<holo_hash::hash_type::Entry>, holochain_wasmer_guest::WasmError>` because of return type
279 | /         hash_entry(Entry::App(AppEntryBytes(
280 | |             SerializedBytes::try_from(self).map_err(|e| wasm_error!(e.into()))?,
281 | |         )))
    | |___________^ expected struct `holo_hash::HoloHash`, found struct `hdi::prelude::HoloHash`
    |
    = note: expected enum `Result<holo_hash::HoloHash<holo_hash::hash_type::Entry>, _>`
               found enum `Result<hdi::prelude::HoloHash<hdi::prelude::holo_hash::hash_type::Entry>, _>`
    = note: perhaps two different versions of crate `holo_hash` are being used?

error[E0277]: the trait bound `holochain_zome_types::LinkTypeFilter: hdi::prelude::LinkTypeFilterExt` is not satisfied
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/hash_path/path.rs:327:26
    |
327 |             let exists = get_links(
    |                          ^^^^^^^^^ the trait `hdi::prelude::LinkTypeFilterExt` is not implemented for `holochain_zome_types::LinkTypeFilter`
    |
    = help: the following other types implement trait `hdi::prelude::LinkTypeFilterExt`:
              &[T; N]
              &[T]
              RangeFull
              Vec<T>
              [T; N]
              hdi::prelude::LinkTypeFilter
note: required by a bound in `link::get_links`
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/link.rs:144:21
    |
142 | pub fn get_links(
    |        --------- required by a bound in this
143 |     base: impl Into<AnyLinkableHash>,
144 |     link_type: impl LinkTypeFilterExt,
    |                     ^^^^^^^^^^^^^^^^^ required by this bound in `link::get_links`

error[E0277]: the trait bound `holochain_zome_types::LinkTypeFilter: hdi::prelude::LinkTypeFilterExt` is not satisfied
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/hash_path/path.rs:340:26
    |
340 |             let exists = get_links(
    |                          ^^^^^^^^^ the trait `hdi::prelude::LinkTypeFilterExt` is not implemented for `holochain_zome_types::LinkTypeFilter`
    |
    = help: the following other types implement trait `hdi::prelude::LinkTypeFilterExt`:
              &[T; N]
              &[T]
              RangeFull
              Vec<T>
              [T; N]
              hdi::prelude::LinkTypeFilter
note: required by a bound in `link::get_links`
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/link.rs:144:21
    |
142 | pub fn get_links(
    |        --------- required by a bound in this
143 |     base: impl Into<AnyLinkableHash>,
144 |     link_type: impl LinkTypeFilterExt,
    |                     ^^^^^^^^^^^^^^^^^ required by this bound in `link::get_links`

error[E0277]: the trait bound `holochain_zome_types::LinkTypeFilter: hdi::prelude::LinkTypeFilterExt` is not satisfied
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/hash_path/path.rs:389:29
    |
389 |         let mut unwrapped = get_links(
    |                             ^^^^^^^^^ the trait `hdi::prelude::LinkTypeFilterExt` is not implemented for `holochain_zome_types::LinkTypeFilter`
    |
    = help: the following other types implement trait `hdi::prelude::LinkTypeFilterExt`:
              &[T; N]
              &[T]
              RangeFull
              Vec<T>
              [T; N]
              hdi::prelude::LinkTypeFilter
note: required by a bound in `link::get_links`
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/link.rs:144:21
    |
142 | pub fn get_links(
    |        --------- required by a bound in this
143 |     base: impl Into<AnyLinkableHash>,
144 |     link_type: impl LinkTypeFilterExt,
    |                     ^^^^^^^^^^^^^^^^^ required by this bound in `link::get_links`

error[E0277]: the trait bound `holochain_zome_types::LinkTypeFilter: hdi::prelude::LinkTypeFilterExt` is not satisfied
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/hash_path/path.rs:438:9
    |
438 |         get_link_details(
    |         ^^^^^^^^^^^^^^^^ the trait `hdi::prelude::LinkTypeFilterExt` is not implemented for `holochain_zome_types::LinkTypeFilter`
    |
    = help: the following other types implement trait `hdi::prelude::LinkTypeFilterExt`:
              &[T; N]
              &[T]
              RangeFull
              Vec<T>
              [T; N]
              hdi::prelude::LinkTypeFilter
note: required by a bound in `link::get_link_details`
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/link.rs:185:21
    |
183 | pub fn get_link_details(
    |        ---------------- required by a bound in this
184 |     base: impl Into<AnyLinkableHash>,
185 |     link_type: impl LinkTypeFilterExt,
    |                     ^^^^^^^^^^^^^^^^^ required by this bound in `link::get_link_details`

error[E0308]: mismatched types
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/link.rs:151:65
    |
151 |                 .get_links(vec![GetLinksInput::new(base.into(), link_type, link_tag)])
    |                                 ------------------              ^^^^^^^^^ expected enum `holochain_zome_types::LinkTypeFilter`, found enum `hdi::prelude::LinkTypeFilter`
    |                                 |
    |                                 arguments to this function are incorrect
    |
    = note: perhaps two different versions of crate `holochain_integrity_types` are being used?
note: associated function defined here
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/holochain_zome_types-0.0.48/src/link.rs:93:12
    |
93  |     pub fn new(
    |            ^^^

error[E0308]: mismatched types
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/link.rs:192:72
    |
192 |                 .get_link_details(vec![GetLinksInput::new(base.into(), link_type, link_tag)])
    |                                        ------------------              ^^^^^^^^^ expected enum `holochain_zome_types::LinkTypeFilter`, found enum `hdi::prelude::LinkTypeFilter`
    |                                        |
    |                                        arguments to this function are incorrect
    |
    = note: perhaps two different versions of crate `holochain_integrity_types` are being used?
note: associated function defined here
   --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/holochain_zome_types-0.0.48/src/link.rs:93:12
    |
93  |     pub fn new(
    |            ^^^

error[E0308]: mismatched types
  --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdk-0.0.152/src/trace.rs:70:24
   |
70 |                   .trace(TraceMsg {
   |  __________________-----_^
   | |                  |
   | |                  arguments to this function are incorrect
71 | |                     level: event.metadata().level().into(),
72 | |                     msg: format!(
73 | |                         "{}:{}:{} {}{}",
...  |
79 | |                     ),
80 | |                 })
   | |_________________^ expected struct `hdi::prelude::TraceMsg`, found struct `holochain_zome_types::TraceMsg`
   |
   = note: perhaps two different versions of crate `holochain_integrity_types` are being used?
note: associated function defined here
  --> /Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/hdi-0.1.3/src/hdi.rs:43:8
   |
43 |     fn trace(&self, trace_msg: TraceMsg) -> ExternResult<()>;
   |        ^^^^^

Some errors have detailed explanations: E0053, E0277, E0308.
For more information about an error, try `rustc --explain E0053`.
error: could not compile `hdk` due to 23 previous errors
```

--- a/crates/hdk/CHANGELOG.md
+++ b/crates/hdk/CHANGELOG.md
@@ -4,6 +4,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 ## Unreleased
 
+* Pin the _hdi_ dependency version. [\#1605](https://github.com/holochain/holochain/pull/1605)
+
 ## 0.0.156
 
 ## 0.0.155

--- a/crates/hdk/Cargo.toml
+++ b/crates/hdk/Cargo.toml
@@ -23,7 +23,7 @@ test_utils = [
 properties = ["holochain_zome_types/properties"]
 
 [dependencies]
-hdi = { version = "0.1.5", path = "../hdi", features = ["trace"] }
+hdi = { version = "=0.1.5", path = "../hdi", features = ["trace"] }
 hdk_derive = { version = "0.0.51", path = "../hdk_derive" }
 holo_hash = { version = "0.0.34", path = "../holo_hash" }
 holochain_wasmer_guest = "=0.0.80"

--- a/crates/release-automation/src/lib/common.rs
+++ b/crates/release-automation/src/lib/common.rs
@@ -56,55 +56,10 @@ pub(crate) fn set_version<'a>(
         .to_owned();
 
     for dependant in dependants.iter() {
-        let target_manifest = dependant.manifest_path();
-
-        debug!(
-            "[{}] updating dependency version from dependant {} to version requirement {} in manifest {:?}",
-            crt.name(),
-            dependant.name(),
-            &version_req,
-            &target_manifest,
-        );
-
-        if !dry_run {
-            set_dependency_version(target_manifest, &crt.name(), &version_req)?;
-        }
+        dependant.set_dependency_version(&crt.name(), &version_req, dry_run)?;
     }
 
     Ok(dependants)
-}
-
-// Adapted from https://github.com/sunng87/cargo-release/blob/f94938c3f20ef20bc8f971d59de75574a0b18931/src/cargo.rs#L122-L154
-fn set_dependency_version(manifest_path: &Path, name: &str, version: &str) -> Fallible<()> {
-    let temp_manifest_path = manifest_path
-        .parent()
-        .ok_or_else(|| anyhow::anyhow!("couldn't get parent of path {}", manifest_path.display()))?
-        .join("Cargo.toml.work");
-
-    {
-        let manifest = load_from_file(manifest_path)?;
-        let mut manifest: toml_edit::Document = manifest.parse()?;
-        for key in &["dependencies", "dev-dependencies", "build-dependencies"] {
-            if manifest.as_table().contains_key(key)
-                && manifest[key]
-                    .as_table()
-                    .expect("manifest is already verified")
-                    .contains_key(name)
-            {
-                let existing_version = manifest[key][name]["version"].as_str().unwrap_or("*");
-
-                if *key == "dependencies" || !existing_version.contains("*") {
-                    manifest[key][name]["version"] = toml_edit::value(version);
-                }
-            }
-        }
-
-        let mut file_out = std::fs::File::create(&temp_manifest_path)?;
-        file_out.write_all(manifest.to_string_in_original_order().as_bytes())?;
-    }
-    std::fs::rename(temp_manifest_path, manifest_path)?;
-
-    Ok(())
 }
 
 #[cfg(test)]
@@ -134,7 +89,8 @@ pub(crate) fn get_dependency_version(manifest_path: &Path, name: &str) -> Fallib
     bail!("version not found")
 }
 
-fn load_from_file(path: &Path) -> Fallible<String> {
+/// Load a file into a String
+pub(crate) fn load_from_file(path: &Path) -> Fallible<String> {
     let mut file = std::fs::File::open(path)?;
     let mut s = String::new();
     file.read_to_string(&mut s)?;

--- a/crates/release-automation/src/lib/common.rs
+++ b/crates/release-automation/src/lib/common.rs
@@ -4,7 +4,7 @@ use std::{
 };
 
 use cargo::util::VersionExt;
-use semver::VersionReq;
+use semver::{Comparator, VersionReq};
 use serde::{Deserialize, Serialize};
 
 use super::*;
@@ -26,40 +26,6 @@ pub(crate) fn selection_check<'a>(
     );
 
     Ok(release_selection)
-}
-
-/// Sets the new version for the given crate and returns the dependants of the crates for post-processing them.
-pub(crate) fn set_version<'a>(
-    dry_run: bool,
-    crt: &'a crate_selection::Crate<'a>,
-    release_version: &semver::Version,
-) -> Fallible<Vec<&'a crate_selection::Crate<'a>>> {
-    let cargo_toml_path = crt.root().join("Cargo.toml");
-    debug!(
-        "setting version to {} in manifest at {:?}",
-        release_version, cargo_toml_path
-    );
-
-    let version = release_version.to_string();
-
-    if !dry_run {
-        cargo_next::set_version(&cargo_toml_path, version.as_str())?;
-    }
-
-    let version_req = format!("={}", release_version);
-
-    let dependants = crt
-        .dependants_in_workspace_filtered(|(_dep_name, deps)| {
-            deps.iter()
-                .any(|dep| dep.version_req() != &cargo::util::OptVersionReq::from(VersionReq::STAR))
-        })?
-        .to_owned();
-
-    for dependant in dependants.iter() {
-        dependant.set_dependency_version(&crt.name(), &version_req, dry_run)?;
-    }
-
-    Ok(dependants)
 }
 
 #[cfg(test)]

--- a/crates/release-automation/src/lib/common.rs
+++ b/crates/release-automation/src/lib/common.rs
@@ -39,9 +39,14 @@ pub(crate) fn set_version<'a>(
         "setting version to {} in manifest at {:?}",
         release_version, cargo_toml_path
     );
+
+    let version = release_version.to_string();
+
     if !dry_run {
-        cargo_next::set_version(&cargo_toml_path, release_version.to_string())?;
+        cargo_next::set_version(&cargo_toml_path, version.as_str())?;
     }
+
+    let version_req = format!("={}", release_version);
 
     let dependants = crt
         .dependants_in_workspace_filtered(|(_dep_name, deps)| {
@@ -54,19 +59,15 @@ pub(crate) fn set_version<'a>(
         let target_manifest = dependant.manifest_path();
 
         debug!(
-            "[{}] updating dependency version from dependant {} to version {} in manifest {:?}",
+            "[{}] updating dependency version from dependant {} to version requirement {} in manifest {:?}",
             crt.name(),
             dependant.name(),
-            release_version.to_string().as_str(),
+            &version_req,
             &target_manifest,
         );
 
         if !dry_run {
-            set_dependency_version(
-                target_manifest,
-                &crt.name(),
-                release_version.to_string().as_str(),
-            )?;
+            set_dependency_version(target_manifest, &crt.name(), &version_req)?;
         }
     }
 

--- a/crates/release-automation/src/lib/crate_.rs
+++ b/crates/release-automation/src/lib/crate_.rs
@@ -4,7 +4,7 @@ use cargo::util::VersionExt;
 use linked_hash_map::LinkedHashMap;
 use linked_hash_set::LinkedHashSet;
 use log::{debug, info, trace, warn};
-use semver::Version;
+use semver::{Comparator, Version, VersionReq};
 use std::collections::{HashMap, HashSet};
 use structopt::StructOpt;
 
@@ -176,7 +176,7 @@ pub(crate) fn cmd(args: &crate::cli::Args, cmd_args: &CrateArgs) -> CommandResul
                 .find(|crt| crt.name() == subcmd_args.crate_name)
                 .ok_or_else(|| anyhow::anyhow!("crate {} not found", subcmd_args.crate_name))?;
 
-            crate::common::set_version(false, crt, &subcmd_args.new_version)?;
+            crt.set_version(false, &subcmd_args.new_version)?;
 
             Ok(())
         }
@@ -222,8 +222,8 @@ pub(crate) fn cmd(args: &crate::cli::Args, cmd_args: &CrateArgs) -> CommandResul
 }
 
 fn pin_deps<'a>(
-    ws: &'a ReleaseWorkspace<'a>,
-    subcmd_args: &CratePinDepsArgs,
+    _ws: &'a ReleaseWorkspace<'a>,
+    _subcmd_args: &CratePinDepsArgs,
 ) -> Result<(), anyhow::Error> {
     todo!()
 }
@@ -244,7 +244,12 @@ fn make_pinned_dep<'a>(
     for dependant in crt.dependants_in_workspace()? {
         dependant.set_dependency_version(
             &crt.name(),
-            &format!("{}{}", subcmd_args.version_prefix, crt.version()),
+            &crt.version(),
+            Some(&semver::VersionReq::parse(&format!(
+                "{}{}",
+                subcmd_args.version_prefix,
+                crt.version()
+            ))?),
             subcmd_args.dry_run,
         )?;
     }
@@ -471,7 +476,7 @@ pub(crate) fn apply_dev_vesrions_to_selection<'a>(
             version,
         );
 
-        for changed_dependant in crate::common::set_version(dry_run, crt, &version)? {
+        for changed_dependant in crt.set_version(dry_run, &version)? {
             if applicable_crates
                 .insert(changed_dependant.name(), changed_dependant)
                 .is_none()

--- a/crates/release-automation/src/lib/crate_.rs
+++ b/crates/release-automation/src/lib/crate_.rs
@@ -123,6 +123,28 @@ pub(crate) struct EnsureCrateOwnersArgs {
 }
 
 #[derive(Debug, StructOpt)]
+pub(crate) struct CratePinDepsArgs {
+    #[structopt(long)]
+    dry_run: bool,
+
+    #[structopt(long, default_value = "=")]
+    version_prefix: String,
+
+    crt: String,
+}
+
+#[derive(Debug, StructOpt)]
+pub(crate) struct CrateMakePinnedArgs {
+    #[structopt(long)]
+    dry_run: bool,
+
+    #[structopt(long, default_value = "=")]
+    version_prefix: String,
+
+    crt: String,
+}
+
+#[derive(Debug, StructOpt)]
 pub(crate) enum CrateCommands {
     SetVersion(CrateSetVersionArgs),
     ApplyDevVersions(CrateApplyDevVersionsArgs),
@@ -135,6 +157,12 @@ pub(crate) enum CrateCommands {
 
     Check(CrateCheckArgs),
     EnsureCrateOwners(EnsureCrateOwnersArgs),
+
+    /// Pins all dependencies of a given crate and its path dependencies recursively
+    PinDeps(CratePinDepsArgs),
+
+    /// Makes a given crate a pinned dependency in the entire workspace
+    MakePinnedDep(CrateMakePinnedArgs),
 }
 
 pub(crate) fn cmd(args: &crate::cli::Args, cmd_args: &CrateArgs) -> CommandResult {
@@ -188,7 +216,40 @@ pub(crate) fn cmd(args: &crate::cli::Args, cmd_args: &CrateArgs) -> CommandResul
         CrateCommands::DetectMissingReleaseheadings(subcmd_args) => {
             cmd_detect_missing_releaseheadings(&ws, subcmd_args)
         }
+        CrateCommands::PinDeps(subcmd_args) => pin_deps(&ws, subcmd_args),
+        CrateCommands::MakePinnedDep(subcmd_args) => make_pinned_dep(&ws, subcmd_args),
     }
+}
+
+fn pin_deps<'a>(
+    ws: &'a ReleaseWorkspace<'a>,
+    subcmd_args: &CratePinDepsArgs,
+) -> Result<(), anyhow::Error> {
+    todo!()
+}
+
+fn make_pinned_dep<'a>(
+    ws: &'a ReleaseWorkspace<'a>,
+    subcmd_args: &CrateMakePinnedArgs,
+) -> Result<(), anyhow::Error> {
+    let crt = ws
+        .members()?
+        .into_iter()
+        .find(|member| member.name() == subcmd_args.crt)
+        .ok_or(anyhow::anyhow!(
+            "looking for crate {} in workspace",
+            subcmd_args.crt
+        ))?;
+
+    for dependant in crt.dependants_in_workspace()? {
+        dependant.set_dependency_version(
+            &crt.name(),
+            &format!("{}{}", subcmd_args.version_prefix, crt.version()),
+            subcmd_args.dry_run,
+        )?;
+    }
+
+    Ok(())
 }
 
 fn cmd_detect_missing_releaseheadings<'a>(

--- a/crates/release-automation/src/lib/crate_selection/mod.rs
+++ b/crates/release-automation/src/lib/crate_selection/mod.rs
@@ -15,7 +15,7 @@ use linked_hash_map::LinkedHashMap;
 use linked_hash_set::LinkedHashSet;
 use once_cell::unsync::{Lazy, OnceCell};
 use regex::Regex;
-use semver::Version;
+use semver::{Comparator, Op, Version, VersionReq};
 use std::cell::Cell;
 use std::collections::{BTreeSet, HashMap, HashSet};
 use std::fmt;
@@ -81,16 +81,51 @@ impl<'a> Crate<'a> {
         self.package.manifest_path()
     }
 
+    /// Sets the new version for the given crate, updates all workspace dependants,
+    /// and returns a refrence to them for post-processing.
+    pub(crate) fn set_version(
+        &'a self,
+        dry_run: bool,
+        release_version: &semver::Version,
+    ) -> Fallible<Vec<&'a Crate<'a>>> {
+        debug!(
+            "setting version to {} in manifest at {:?}",
+            release_version,
+            self.manifest_path(),
+        );
+
+        let release_version_str = release_version.to_string();
+
+        if !dry_run {
+            cargo_next::set_version(self.manifest_path(), release_version_str.as_str())?;
+        }
+
+        let dependants = self
+            .dependants_in_workspace_filtered(|(_dep_name, deps)| {
+                deps.iter().any(|dep| {
+                    dep.version_req() != &cargo::util::OptVersionReq::from(VersionReq::STAR)
+                })
+            })?
+            .to_owned();
+
+        for dependant in dependants.iter() {
+            dependant.set_dependency_version(&self.name(), &release_version, None, dry_run)?;
+        }
+
+        Ok(dependants)
+    }
+
     /// Set a dependency to a specific version
     // Adapted from https://github.com/sunng87/cargo-release/blob/f94938c3f20ef20bc8f971d59de75574a0b18931/src/cargo.rs#L122-L154
     pub(crate) fn set_dependency_version(
         &self,
         name: &str,
-        version: &str,
+        version: &Version,
+        version_req_override: Option<&VersionReq>,
         dry_run: bool,
     ) -> Fallible<()> {
         debug!(
-            "[{}] updating dependency version from dependant {} to version requirement {} in manifest {:?}",
+            "[{}] updating dependency version from dependant {} to version {} in manifest {:?}",
             &self.name(),
             &name,
             &version,
@@ -118,10 +153,69 @@ impl<'a> Crate<'a> {
                         .expect("manifest is already verified")
                         .contains_key(name)
                 {
-                    let existing_version = manifest[key][name]["version"].as_str().unwrap_or("*");
+                    let existing_version_req = if let Some(Ok(existing_version_req)) =
+                        manifest[key][name]["version"].as_str().map(|version| {
+                            VersionReq::parse(version).context(anyhow::anyhow!(
+                                "parsing version {:?} for dependency {} ",
+                                version,
+                                self.name()
+                            ))
+                        }) {
+                        existing_version_req
+                    } else {
+                        debug!(
+                            "could not parse {}'s {} version req to string: {:?}",
+                            name, key, manifest[key][name]["version"]
+                        );
 
-                    if *key == "dependencies" || !existing_version.contains("*") {
-                        manifest[key][name]["version"] = toml_edit::value(version);
+                        continue;
+                    };
+
+                    trace!(
+                        "version: {:?}, existing version req {:?} in {}",
+                        version,
+                        existing_version_req,
+                        key,
+                    );
+
+                    // only set the version if necessary
+                    if *key == "dependencies" || existing_version_req != VersionReq::STAR {
+                        let final_version_req = if let Some(vr) = version_req_override {
+                            vr.clone()
+                        } else {
+                            let mut version_req = VersionReq::parse(&version.to_string())?;
+
+                            // if the Op of the first Comparator we'll inherit that, the rest will be discarded
+                            if let Some(op) = existing_version_req
+                                .comparators
+                                .first()
+                                .map(|comp| {
+                                    if comp.op != semver::Op::Wildcard {
+                                        Some(comp.op)
+                                    } else {
+                                        None
+                                    }
+                                })
+                                .flatten()
+                            {
+                                trace!("overriding first op of {:?} with {:?}", version_req, op);
+                                let version_req_clone = version_req.clone();
+
+                                version_req
+                                    .comparators
+                                    .first_mut()
+                                    .ok_or_else(|| anyhow::anyhow!{
+                                        "first comparator of version_req {:?} should be accessible",
+                                        version_req_clone
+                                    })?
+                                    .op = op;
+                            };
+
+                            version_req
+                        };
+
+                        manifest[key][name]["version"] =
+                            toml_edit::value(final_version_req.to_string());
                     }
                 }
             }

--- a/crates/release-automation/src/lib/crate_selection/mod.rs
+++ b/crates/release-automation/src/lib/crate_selection/mod.rs
@@ -19,6 +19,7 @@ use semver::Version;
 use std::cell::Cell;
 use std::collections::{BTreeSet, HashMap, HashSet};
 use std::fmt;
+use std::io::Write;
 use std::iter::FromIterator;
 use std::path::{Path, PathBuf};
 use std::process::Command;
@@ -78,6 +79,61 @@ impl<'a> Crate<'a> {
     /// Return the path of the package's manifest.
     pub(crate) fn manifest_path(&self) -> &Path {
         self.package.manifest_path()
+    }
+
+    /// Set a dependency to a specific version
+    // Adapted from https://github.com/sunng87/cargo-release/blob/f94938c3f20ef20bc8f971d59de75574a0b18931/src/cargo.rs#L122-L154
+    pub(crate) fn set_dependency_version(
+        &self,
+        name: &str,
+        version: &str,
+        dry_run: bool,
+    ) -> Fallible<()> {
+        debug!(
+            "[{}] updating dependency version from dependant {} to version requirement {} in manifest {:?}",
+            &self.name(),
+            &name,
+            &version,
+            self.manifest_path(),
+        );
+
+        let temp_manifest_path = self
+            .manifest_path()
+            .parent()
+            .ok_or_else(|| {
+                anyhow::anyhow!(
+                    "couldn't get parent of path {}",
+                    self.manifest_path().display()
+                )
+            })?
+            .join("Cargo.toml.work");
+
+        {
+            let manifest = crate::common::load_from_file(self.manifest_path())?;
+            let mut manifest: toml_edit::Document = manifest.parse()?;
+            for key in &["dependencies", "dev-dependencies", "build-dependencies"] {
+                if manifest.as_table().contains_key(key)
+                    && manifest[key]
+                        .as_table()
+                        .expect("manifest is already verified")
+                        .contains_key(name)
+                {
+                    let existing_version = manifest[key][name]["version"].as_str().unwrap_or("*");
+
+                    if *key == "dependencies" || !existing_version.contains("*") {
+                        manifest[key][name]["version"] = toml_edit::value(version);
+                    }
+                }
+            }
+
+            let mut file_out = std::fs::File::create(&temp_manifest_path)?;
+            file_out.write_all(manifest.to_string_in_original_order().as_bytes())?;
+        }
+        if !dry_run {
+            std::fs::rename(temp_manifest_path, self.manifest_path())?;
+        }
+
+        Ok(())
     }
 
     /// Return a reference to the package.
@@ -1082,8 +1138,15 @@ impl<'a> ReleaseWorkspace<'a> {
                     vec![
                         vec!["fetch", "--verbose", "--manifest-path", mp],
                         vec![
-                            vec!["update", "--workspace", "--offline", "--verbose", "--manifest-path", mp],
-                            if dry_run { vec!["--dry-run"] } else { vec![] }
+                            vec![
+                                "update",
+                                "--workspace",
+                                "--offline",
+                                "--verbose",
+                                "--manifest-path",
+                                mp,
+                            ],
+                            if dry_run { vec!["--dry-run"] } else { vec![] },
                         ]
                         .concat(),
                     ]

--- a/crates/release-automation/src/lib/release.rs
+++ b/crates/release-automation/src/lib/release.rs
@@ -243,7 +243,7 @@ fn bump_release_versions<'a>(
 
         let greater_release = release_version > current_version;
         if greater_release {
-            common::set_version(cmd_args.dry_run, crt, &release_version.clone())?;
+            crt.set_version(cmd_args.dry_run, &release_version.clone())?;
         }
 
         let crate_release_heading_name = format!("{}", release_version);

--- a/crates/release-automation/src/lib/tests/cli.rs
+++ b/crates/release-automation/src/lib/tests/cli.rs
@@ -149,7 +149,7 @@ fn bump_versions_on_selection() {
     // ensure dependants were updated
     // todo: ensure *all* dependants were updated
     assert_eq!(
-        "0.0.0",
+        "=0.0.0",
         crate::common::get_dependency_version(
             &workspace
                 .root()

--- a/crates/release-automation/src/lib/tests/cli.rs
+++ b/crates/release-automation/src/lib/tests/cli.rs
@@ -146,23 +146,25 @@ fn bump_versions_on_selection() {
         get_crate_versions(&expected_crates, &workspace),
     );
 
-    // ensure dependants were updated
-    // todo: ensure *all* dependants were updated
-    assert_eq!(
-        "=0.0.0",
-        crate::common::get_dependency_version(
-            &workspace
-                .root()
-                .join("crates")
-                .join("crate_a")
-                .join("Cargo.toml"),
-            "crate_b"
-        )
-        .unwrap()
-        .replace("\"", "")
-        .replace("\\", "")
-        .replace(" ", ""),
-    );
+    // ensure *all* dependants were updated
+    // alas, after refactoring the code into a loop i noticed there's only one dependency in this example workspace
+    for (name, dep_name, expected_crate_version) in &[("crate_a", "crate_b", "=0.0.0")] {
+        assert_eq!(
+            expected_crate_version,
+            &crate::common::get_dependency_version(
+                &workspace
+                    .root()
+                    .join("crates")
+                    .join(name)
+                    .join("Cargo.toml"),
+                dep_name
+            )
+            .unwrap()
+            .replace("\"", "")
+            .replace("\\", "")
+            .replace(" ", ""),
+        );
+    }
 
     // check changelogs for new release headings
     assert_eq!(

--- a/crates/release-automation/src/lib/tests/workspace_mocker.rs
+++ b/crates/release-automation/src/lib/tests/workspace_mocker.rs
@@ -328,7 +328,7 @@ pub(crate) fn example_workspace_1<'a>() -> Fallible<WorkspaceMocker> {
             name: "crate_a".to_string(),
             version: "0.0.1".to_string(),
             dependencies: vec![
-                r#"crate_b = { path = "../crate_b", version = "0.0.0-alpha.1" }"#.to_string(),
+                r#"crate_b = { path = "../crate_b", version = "=0.0.0-alpha.1" }"#.to_string(),
             ],
             excluded: false,
             ty: workspace_mocker::MockProjectType::Bin,

